[formrecognizer] Add migration guide (#20960)
Azure/azure-sdk-for-python: sdk/formrecognizer/azure-ai-formrecognizer/MIGRATION_GUIDE.md

> # Guide for migrating azure-ai-formrecognizer to version 3.2.x from versions 3.1.x and below

Author: would appreciate any feedback on additional info that should be included

> This guide is intended to assist in the migration to `azure-ai-formrecognizer (3.2.x)` from versions `3.1.x` and below. It will focus on side-by-side comparisons for similar operations between versions, please note that version `3.1.2` will be used for comparison with `3.2.0b1`.

Author: Yeah I think that's what I did here? If I understood you correctly?

> This guide is intended to assist in the migration to `azure-ai-formrecognizer (3.2.x)` from versions `3.1.x` and below. It will focus on side-by-side comparisons for similar operations between versions, please note that version `3.1.2` will be used for comparison with `3.2.0b1`.

Author: Ah you're talking about the last mention of these version, will update

Familiariy with `azure-ai-formrecognizer (3.1.x and below)` package is assumed. For those new to the Azure Form Recognizer SDK please refer to the [README](README) rather than this guide.

## Table of Contents
- [Migration benefits](#migration-benefits)
- [Important changes](#important-changes)
- [Additional samples](#additional-samples)

## Migration benefits

A natural question to ask when considering whether or not to adopt a new version or library is what the benefits of doing so would be. As Azure Form Recognizer has matured and been embraced by a more diverse group of developers, we have been focused on learning the patterns and practices to best support developer productivity and add value to our customers.

The benefit of the new design of the `azure-ai-formrecognizer (3.2.x)` library is that it introduces two new clients `DocumentAnalysisClient` and the `DocumentModelAdministrationClient` with unified methods for analyzing documents and provides support for the new features added by the service in version `2021-09-30-preview`.
Please refer to the [README](README) for more information on these new clients. 

## Important changes

> ### Instantiating a client

Reviewer: maybe the title should be titled "Clients usage" or something like that?

Author: @kristapratico thoughts? I'm fine with either, I would put `Client usage`.

Reviewer 2: Either seems fine to me

We continue to support connection strings and AAD authentication methods when creating our clients. Below are the differences between the two versions:

- In `3.2.x`, we have added `DocumentAnalysisClient` and `DocumentModelAdministrationClient` which support service version `2021-09-30-preview`.
- `FormRecognizerClient` and `FormTrainingClient` will return an error is called with an API version of `2021-09-30-preview` or above. 

Creating a new client in `3.1.x`:
```python
form_recognizer_client = FormRecognizerClient(
    endpoint=endpoint, credential=AzureKeyCredential(key)
)
```

Creating a new client in `3.2.x`:
```python
document_analysis_client = DocumentAnalysisClient(
    endpoint=endpoint, credential=AzureKeyCredential(key)
)
```

> ### Analyzing documents

Reviewer: it seems like there is too much here. I would prefer a specific section for prebuilt old vs new (receipt, businesscard, etc),
then another for recognizecontent.
the prebuilt-document is a new thing so it doesn;t have an equivalent before so no need to add it to this guide

Author: Reorganized a bit so that there isnt as much there. I'll think about separating into prebuilt and other content, adding an issue here: https://github.com/Azure/azure-sdk-for-python/issues/21001

Differences between the versions:

> - In `DocumentAnalysisClient` all prebuilt model methods along with custom model, layout and general prebuilt document analysis are unified into two methods called

Reviewer: ```suggestion
- In `DocumentAnalysisClient` all prebuilt model methods along with custom model, layout, and prebuilt document analysis are unified into two methods called
```

> `begin_analyze_document` and `begin_analyze_document_from_url`.

Reviewer 2: maybe describe how `model` is passed in to choose the type of analysis?

- In `FormRecognizerClient` there are two methods (a stream and URL method) for each of the prebuilt models supported by the service, this results in two methods for business card, receipt, identity document, and invoice models, along with a pair of methods for recognizing custom documents and for recognizing content/layout. 
- Along with more consolidated analysis methods in the `DocumentAnalysisClient`, the return types have also been improved and remove the hierarchical dependencies between elements. An instance of the `AnalyzeResult` model is now returned which moves important document elements to the top level of the returned model.

> - The `include_field_elements` kwarg does not exist with the `DocumentAnalysisClient`, text details are automatically included with service version `2021-09-30-preview`.

Author: This is repeated down below with custom models, I dont like that it's repetitive but I also felt it was important to let the user know in both places 😆 let me know which one I should keep

> - Along with more consolidated analysis methods in the `DocumentAnalysisClient`, the return types have also been improved and remove the hierarchical dependencies between elements. An instance of the `AnalyzeResult` model is now returned which moves important document elements to the top level of the returned model.
> - The `include_field_elements` kwarg does not exist with the `DocumentAnalysisClient`, text details are automatically included with service version `2021-09-30-preview`.

Reviewer 2: I think we should stick with one - service or API version

```suggestion
- Along with more consolidated analysis methods in the `DocumentAnalysisClient`, the return types have also been improved and remove the hierarchical dependencies between elements. An instance of the `AnalyzeResult` model is now returned which moves important document elements to the top level of the returned model.
- The `include_field_elements` kwarg does not exist with the `DocumentAnalysisClient`, text details are automatically included with API version `2021-09-30-preview`.
```

> - The `include_field_elements` kwarg does not exist with the `DocumentAnalysisClient`, text details are automatically included with service version `2021-09-30-preview`.

Reviewer 2: might be worth mentioning in a bullet point that content -> layout with the new version

> - The `include_field_elements` kwarg does not exist with the `DocumentAnalysisClient`, text details are automatically included with service version `2021-09-30-preview`.

Author: I was referring to the fact that things arent nested in a form page anymore...like tables, entities, kv pairs, etc

> - The `include_field_elements` kwarg does not exist with the `DocumentAnalysisClient`, text details are automatically included with service version `2021-09-30-preview`.

Author: What about this: `An instance of the `AnalyzeResult` model is now returned which showcases important document elements, such as key-value pairs, entities, and documents, among others at the top level of the returned model.`

> - The `include_field_elements` kwarg does not exist with the `DocumentAnalysisClient`, text details are automatically included with service version `2021-09-30-preview`.

Reviewer 2: Love it! I would add "tables" to that list too and maybe say "document fields and values" instead of just "documents"

Analyzing prebuilt models like business cards, identity documents, invoices and receipts with `3.1.x`:
```python
with open(path_to_sample_forms, "rb") as f:
    poller = form_recognizer_client.begin_recognize_receipts(receipt=f, locale="en-US")
receipts = poller.result()

for idx, receipt in enumerate(receipts):
    print("--------Recognizing receipt #{}--------".format(idx+1))

>     # process receipt

Author: I was on the fence about including the fields...what do you think? I was worried that the lengthy sample code would overpower the main point of the differences...but I can see an argument for why it's better to include the fields to showcase using the result in the different versions.

>     # process receipt

Reviewer 2: maybe just include one or two fields? Not much code in python and I think it's important to have runnable snippets throughout.

```

Analyzing prebuilt models like business cards, identity documents, invoices and receipts with `3.2.x`:
```python
with open(path_to_sample_documents, "rb") as f:
    poller = document_analysis_client.begin_analyze_document(
        "prebuilt-receipt", document=f, locale="en-US"
    )
receipts = poller.result()

for idx, receipt in enumerate(receipts.documents):
    print("--------Recognizing receipt #{}--------".format(idx + 1))
    # process receipt
```

Analyzing document content with `3.1.x`:
```python
with open(path_to_sample_forms, "rb") as f:

>     poller = form_recognizer_client.begin_recognize_content(form=f)

Author: Putting it in-line here with the samples....but I could see if being a part of the main list of bullet point of changes.

form_pages = poller.result()

for idx, content in enumerate(form_pages):
    print("----Recognizing content from page #{}----".format(idx+1))
    # process document layout by parsing content in pages

>     for table_idx, table in enumerate(content.tables):

Author: I copy pasted this from the v3.1 samples, should I still remove it?

>     for table_idx, table in enumerate(content.tables):

Author: I think this would depend on whether or not I include the full sample code, which is related to a previous comment. Let me know!

>     for table_idx, table in enumerate(content.tables):

Author: Ok going to keep some of the instances of enumerate since I'm going to be using a fuller version of our samples.

```python
with open(path_to_sample_documents, "rb") as f:
    poller = document_analysis_client.begin_analyze_document(
        "prebuilt-layout", document=f
    )
result = poller.result()

for idx, style in enumerate(result.styles):
    # process document styles

for page in result.pages:
    # process pages
    for line in page.lines:
        # process lines
    for word in page.words:
        # process words

for table_idx, table in enumerate(result.tables):
    # process tables found in document
```

Analyzing general prebuilt document types with `3.2.x`:
```python
with open(path_to_sample_documents, "rb") as f:
    poller = document_analysis_client.begin_analyze_document(
        "prebuilt-document", document=f
    )
result = poller.result()

for entity in result.entities:
    # process entities found in document

for kv_pair in result.key_value_pairs:
    # process key-value pairs found in document

for page in result.pages:
    # process pages
    for line in page.lines:
        # process lines
    for word in page.words:
        # process words

for table_idx, table in enumerate(result.tables):
    # process tables found in document
```

> NOTE: All of these samples also work with `begin_analyze_document_from_url` when providing a valid URL to the document.

Reviewer 2: ```suggestion
> NOTE: All of these samples also work with `begin_analyze_document_from_url` when providing a valid URL to the document.
```

### Analyzing a custom model

Differences between the versions:
- Analyzing a custom model with `DocumentAnalysisClient` uses the general `begin_analyze_document` and `begin_analyze_document_from_url` methods.
- In order to analyze a custom model with `FormRecognizerClient` the `begin_recognize_custom_models` and its corresponding URL methods are used.
- The `include_field_elements` kwarg does not exist with the `DocumentAnalysisClient`, text details are automatically included with service version `2021-09-30-preview`.

Analyze custom document with `3.1.x`:
```python
with open(path_to_sample_forms, "rb") as f:
    poller = form_recognizer_client.begin_recognize_custom_forms(
        model_id=model_id, form=f, include_field_elements=True
    )
forms = poller.result()
```

Analyze custom document with `3.2.x`:
```python
with open(path_to_sample_documents, "rb") as f:
    poller = document_analysis_client.begin_analyze_document(
        model=model_id, document=f
    )
result = poller.result()
```

> ### Training a custom model

Author: FYI that I used training here since this guide is for users who are using the old version and moving to the new one. Within the text I do mention build, but this was on purpose. 

Also, I'm aware that I used analyzing above (I felt that one was not to far a leap from recognizing), but let me know if I should unify terms to those used in a certain version

Differences between the versions:

> - Files for building a new model for version `3.2.x` can be created using the new labeling tool found [here](TODO).

Reviewer: I wouldn't say the tool is new. maybe a new version of the tool?

> - Files for building a new model for version `3.2.x` can be created using the new labeling tool found [here](TODO).

Reviewer: I wouldn't say the tool is new. maybe a new version of the tool?

> - Files for building a new model for version `3.2.x` can be created using the new labeling tool found [here](TODO).

Author: Removed the reference to new.

> - In version `3.1.x` the `use_training_labels` kwarg was used to indicate whether custom labeling was when creating the new models.

Reviewer 2: ```suggestion
- In version `3.1.x` the `use_training_labels` kwarg was used to indicate whether to use labeled data when creating the custom model.
```

> - In version `3.2.x` the `use_training_labels` kwargs is not supported since training must be carried out with labeled training documents. In order to extract key-value pairs from a document, please refer to the prebuilt model "prebuilt-document" which extracts entities, key-value pairs, and layout from a document. 

Reviewer 2: Since we didn't expose key-value pairs as key value pairs in v3.1.x, you might want to describe how prebuilt-document satisfies the training without labels use case + the new stuff it returns.

> 

Reviewer 2: I think we should add how you can use `model_id` to actually name the model in the new version, not just add a display/model_name like before

Reviewer 2: Could also add comment about the differences conceptually - no more submodels, we just have a model that can analyze different doctypes.

Author: Added the model id comment at the top in the benefits, also adding it here.

Train a custom model with `3.1.x`:
```python

> form_training_client = FormTrainingClient(endpoint, AzureKeyCredential(key), api_version="2.1")

Reviewer 2: nit we don't include this in the other samples

```suggestion
form_training_client = FormTrainingClient(endpoint, AzureKeyCredential(key))
```

poller = form_training_client.begin_training(
    container_sas_url, use_training_labels=True, model_name="mymodel"
)
model = poller.result()

# Custom model information
print("Model ID: {}".format(model.model_id))
print("Status: {}".format(model.status))
print("Model name: {}".format(model.model_name))
print("Is this a composed model?: {}".format(model.properties.is_composed_model))
print("Training started on: {}".format(model.training_started_on))
print("Training completed on: {}".format(model.training_completed_on))

print("Recognized fields:")
# looping through the submodels, which contains the fields they were trained on

> # The labels are based on the ones you gave the training document.

Reviewer 2: this comment is specific to the train with labels samples, i'd omit

for submodel in model.submodels:
    print("...The submodel has model ID: {}".format(submodel.model_id))
    print("...The submodel with form type {} has an average accuracy '{}'".format(
        submodel.form_type, submodel.accuracy
    ))
    for name, field in submodel.fields.items():
        print("...The model found the field '{}' with an accuracy of {}".format(
            name, field.accuracy
        ))

# Training result information
for doc in model.training_documents:
    print("Document name: {}".format(doc.name))
    print("Document status: {}".format(doc.status))
    print("Document page count: {}".format(doc.page_count))
    print("Document errors: {}".format(doc.errors))
```

Train a custom model with `3.2.x`:
```python
document_model_admin_client = DocumentModelAdministrationClient(endpoint, AzureKeyCredential(key))
poller = document_model_admin_client.begin_build_model(

>     container_sas_url, description="my model description"

Reviewer 2: I think we should add `model_id` kwarg here

)
model = poller.result()

print("Model ID: {}".format(model.model_id))
print("Description: {}".format(model.description))
print("Model created on: {}\n".format(model.created_on))
print("Doc types the model can recognize:")
for name, doc_type in model.doc_types.items():
    print("\nDoc Type: '{}' which has the following fields:".format(name))

>     for field_name, field in doc_type.field_schema.items():
>         print("Field: '{}' has type '{}' and confidence score {}".format(
>             field_name, field["type"], doc_type.field_confidence[field_name]
>         ))

Reviewer 2: I think we should update to Paul's recommendation here:

```suggestion
    for field_name, confidence in doc_type.field_confidence.items():
        print("Field: '{}' has confidence score {}".format(field_name, confidence))
```

```

> 

Reviewer: Java's PR has a section for manage models that is important to call out too. See https://github.com/Azure/azure-sdk-for-java/pull/24472

Author: So I added the info that Java has there under training a custom model, I could reorganize a bit but I dont feel I have a lot for that section, I could add the samples for listing models I suppose? Functionally those dont change too much though

Reviewer: I guess the information that was important for get and list models is how in previous versions you will get models with any status, and now only models with a succeeded status are included

Author: Ok added it, separated the bullet points into a manage models section, thanks for pointing this out!

> ### Other differences

Author: I dont like this tite, open to suggestions or even moving this up to be part of the list directly under "important changes"

Reviewer 2: what about Additional Improvements?

> - The service has introduced the concept of a `BoundingRegion` which groups elements or documents through the bounding boxes in which they appear per page.

Reviewer 2: I think there is more we can add here about spans, cross-page elements, styles, etc like Paul has here: https://github.com/Azure/azure-rest-api-specs-pr/blob/paulhsu/formrecognizer.v3.0-preview.1/specification/cognitiveservices/data-plane/FormRecognizer/preview/v3.0-preview.1/doc/Migration.md#content--spans

Reviewer: FYI @samvaity

Author: Added a little more info about that at the top and actually removed the Additional Improvements section. Let me know if it needs more info.

Reviewer: I think it looks great!

# Guide for migrating azure-ai-formrecognizer to version 3.2.x from versions 3.1.x and below

This guide is intended to assist in the migration to `azure-ai-formrecognizer (3.2.x)` from versions `3.1.x` and below. It will focus on side-by-side comparisons for similar operations between versions. Please note that version `3.2.0b1` will be used for comparison with `3.1.2`.

Familiarity with `azure-ai-formrecognizer (3.1.x and below)` package is assumed. For those new to the Azure Form Recognizer client library for Python please refer to the [README](README) rather than this guide.

## Table of Contents
- [Migration benefits](#migration-benefits)
- [Important changes](#important-changes)
    - [Client usage](#client-usage)
    - [Analyzing document](#analyzing-documents)
    - [Analyzing a custom model](#analyzing-a-custom-model)

>     - [Training a custom model](#trainig-a-custom-model)

Reviewer: ```suggestion
    - [Training a custom model](#training-a-custom-model)
```

- [Additional samples](#additional-samples)

## Migration benefits

A natural question to ask when considering whether or not to adopt a new version of the library is what the benefits of doing so would be. As Azure Form Recognizer has matured and been embraced by a more diverse group of developers, we have been focused on learning the patterns and practices to best support developer productivity and add value to our customers.

> There are many benefits to using the new design of the `azure-ai-formrecognizer (3.2.x)` library. This new version of the library introduces two new clients `DocumentAnalysisClient` and the `DocumentModelAdministrationClient` with unified methods for analyzing documents and provides support for the new features added by the service in API version `2021-09-30-preview`.

Reviewer 2: Might be worth adding an "and later" everytime we specify the latest api version so this guide doesn't seem quickly out of date

```suggestion
There are many benefits to using the new design of the `azure-ai-formrecognizer (3.2.x)` library. This new version of the library introduces two new clients `DocumentAnalysisClient` and the `DocumentModelAdministrationClient` with unified methods for analyzing documents and provides support for the new features added by the service in API version `2021-09-30-preview` and later.
```

> New features provided by the `DocumentAnalysisClient` include having one consolidated function for analyzing general document layout, having a general prebuilt document model type, along with the same prebuilt model types that we included previously (receipts, invoices, business cards, identity documents), and custom models. Moreover, the models used in the newer version of the library, such as `AnalyzeResult`, remove many hierarchical dependencies between document elements and move them to a more top level and easily accesible position. The service has further improved how to define where elements are found on documents by moving towards `BoundingRegion` definitions where an element is returned with the bounding box per page where it is found. Document element fields are returned with more information, such as content and spans. 

Reviewer 2: nit: some wording, feel free to ignore or take whatever feedback sounds right to you

```suggestion
New features provided by the `DocumentAnalysisClient` include one consolidated function for analyzing document layout, a general prebuilt document model, along with the same prebuilt models that were included previously (receipts, invoices, business cards, identity documents), and custom models. Moreover, the models introduced in the latest version of the library, such as `AnalyzeResult`, remove hierarchical dependencies between document elements and move them to a more top level and easily accessible position. The service has further improved how to define where elements are located on documents by moving towards `BoundingRegion` definitions allowing for cross-page elements. Document element fields are returned with more information, such as content and spans. 
```

> When using the `DocumentModelAdministrationClient` to build or compose new models users can now assign their own model ids and specify a description. Listing models on the administation client includes both prebuilt and custom models. When using `get_model()`, users can get the field schema for the model they specified, this includes prebuilt models. This client also provides functions for getting information from model operations.

Reviewer 2: more nits on wording

```suggestion
When using the `DocumentModelAdministrationClient` to build, compose, or copy models, users can now assign their own model IDs and specify a description. Listing models on the administration client now includes both prebuilt and custom models. When using `get_model()`, users can get the field schema (field names and types that the model can extract) for the model they specified, including for prebuilt models. This client also provides functions for getting information from model operations.
```

This table shows the relationship between SDK versions and supported API versions of the service

|SDK version|Supported API version of service
|-|-
|3.2.0b1 - Latest beta release | 2.0, 2.1, 2021-09-30-preview
|3.1.X - Latest GA release| 2.0, 2.1 (default)

> |3.0.0| 2.0

Reviewer 2: Since we are talking about the clients in this section maybe we should include the client vs api version table instead?

> Please refer to the [README](README) for more information on these new clients.

Reviewer: ```suggestion
Please refer to the [README][README] for more information on these new clients.
```

Author: Done

## Important changes

### Client usage

> We continue to support API key and AAD authentication methods when creating our clients. Below are the differences between the two versions:

Reviewer 2: ```suggestion
We continue to support API key and AAD authentication methods when creating the clients. Below are the differences between the two versions:
```

- In `3.2.x`, we have added `DocumentAnalysisClient` and `DocumentModelAdministrationClient` which support API version `2021-09-30-preview` and later.

> - `FormRecognizerClient` and `FormTrainingClient` will return an error if called with an API version of `2021-09-30-preview` or later. 

Reviewer 2: ```suggestion
- `FormRecognizerClient` and `FormTrainingClient` will raise an error if called with an API version of `2021-09-30-preview` or later. 
```

- In `DocumentAnalysisClient` all prebuilt model methods along with custom model, layout, and a prebuilt document analysis model are unified into two methods called
`begin_analyze_document` and `begin_analyze_document_from_url`.

> - In `FormRecognizerClient` there are two methods (a stream and URL method) for each of the prebuilt models supported by the service, this results in two methods for business card, receipt, identity document, and invoice models, along with a pair of methods for recognizing custom documents and for recognizing content/layout. 

Reviewer: ```suggestion
- In `FormRecognizerClient` there are two methods (a stream and URL method) for each of the prebuilt models supported by the service. This results in two methods for business card, receipt, identity document, and invoice models, along with a pair of methods for recognizing custom documents and for recognizing content/layout. 
```

> Creating a new clients in `3.1.x`:

Reviewer: ```suggestion
Creating new clients in `3.1.x`:
```

Author: I can do either:
`Creating a new client` or leave it as `Creating new clients`. Adding the a would make it singular. Let me know!

Reviewer: plop. English. Creating new clients?

Author: Done!

```python
form_recognizer_client = FormRecognizerClient(
    endpoint=endpoint, credential=AzureKeyCredential(key)
)

form_training_client = FormTrainingClient(
    endpoint=endpoint, credential=AzureKeyCredential(key)
)
```

> Creating a new clients in `3.2.x`:

Reviewer: ```suggestion
Creating new clients in `3.2.x`:
```

Author: Same comment as above.

```python
document_analysis_client = DocumentAnalysisClient(
    endpoint=endpoint, credential=AzureKeyCredential(key)
)

document_model_admin_client = DocumentModelAdministrationClient(
    endpoint=endpoint, credential=AzureKeyCredential(key)
)
```

### Analyzing documents

Differences between the versions:

> - `begin_analyze_document` and `begin_analyze_document_from_url` accept a string with the desired model id for analysis. The model id can be any of the prebuilt model ids or a custom model id.

Reviewer 2: other instances on this page

```suggestion
- `begin_analyze_document` and `begin_analyze_document_from_url` accept a string with the desired model ID for analysis. The model ID can be any of the prebuilt model ids or a custom model ID.
```

- Along with more consolidated analysis methods in the `DocumentAnalysisClient`, the return types have also been improved and remove the hierarchical dependencies between elements. An instance of the `AnalyzeResult` model is now returned which showcases important document elements, such as key-value pairs, entities, tables, and document fields and values, among others, at the top level of the returned model. This can be contrasted with `RecognizedForm` which included more hierarchical relationships, for instance tables were an element of a `FormPage` and not a top-level element.

> - In the new version of the library, the functionality of `begin_recognize_content` has been added as a prebuilt model and can be called in library version `azure-ai-formrecognizer (3.2.x)` with `begin_analyze_document` by passing in the `prebuilt-layout` model id. Similarly, to get general prebuilt document information, such as key-value pairs, entities, and text layout, the `prebuilt-document` model id can be used with `begin_analyze_document`.

Reviewer: should we add that Reading Order is long gone and now it is always using Natural?

Author: Sounds like a good idea, will add.

- When calling `begin_analyze_document` and `begin_analyze_document_from_url` the returned type is an `AnalyzeResult` object, while the various methods used with `FormRecognizerClient` return a list of `RecognizedForm`.

> - The `pages` kwarg is string with library version `azure-ai-formrecognizer (3.2.x)`. In `azure-ai-formrecognizer (3.1.x)`, pages was a list of strings.

Reviewer 2: ```suggestion
- The `pages` keyword argument is a string with library version `azure-ai-formrecognizer (3.2.x)`. In `azure-ai-formrecognizer (3.1.x)`, `pages` was a list of strings.
```

> - The `include_field_elements` kwarg does not exist with the `DocumentAnalysisClient`, text details are automatically included with API version `2021-09-30-preview`.

Reviewer 2: ```suggestion
- The `include_field_elements` keyword argument is not supported with the `DocumentAnalysisClient`, text details are automatically included with API version `2021-09-30-preview` and later.
```

                    "......Individual Item Price: {} has confidence: {}".format(
                        item_price.value, item_price.confidence
                    )
                )
            item_total_price = item.value.get("TotalPrice")
            if item_total_price:
                print(
                    "......Total Item Price: {} has confidence: {}".format(
                        item_total_price.value, item_total_price.confidence
                    )
                )
    subtotal = receipt.fields.get("Subtotal")
    if subtotal:
        print(
            "Subtotal: {} has confidence: {}".format(
                subtotal.value, subtotal.confidence
            )
        )
    tax = receipt.fields.get("Tax")
    if tax:
        print("Tax: {} has confidence: {}".format(tax.value, tax.confidence))
    tip = receipt.fields.get("Tip")
    if tip:
        print("Tip: {} has confidence: {}".format(tip.value, tip.confidence))
    total = receipt.fields.get("Total")
    if total:
        print("Total: {} has confidence: {}".format(total.value, total.confidence))
    print("--------------------------------------")
```

Analyzing document content with `3.1.x`:

> > NOTE: With version `3.1.x` of the library this method was called with a `language` kwarg to hint at the language for the document, whereas in version `3.2.x` of the library `locale` is used for this purpose.

Reviewer 2: sorry last time I'll mention it

```suggestion
> NOTE: With version `3.1.x` of the library this method was called with a `language` keyword argument to hint at the language for the document, whereas in version `3.2.x` of the library `locale` is used for this purpose.
```

                selection_mark.state,
                format_bounding_box(selection_mark.bounding_box),
                selection_mark.confidence,
            )
        )

for table_idx, table in enumerate(result.tables):
    print(
        "Table # {} has {} rows and {} columns".format(
            table_idx, table.row_count, table.column_count
        )
    )
    for region in table.bounding_regions:
        print(
            "Table # {} location on page: {} is {}".format(
                table_idx,
                region.page_number,
                format_bounding_box(region.bounding_box),
            )
        )
    for cell in table.cells:
        print(
            "...Cell[{}][{}] has text '{}'".format(
                cell.row_index,
                cell.column_index,
                cell.content,
            )
        )
        for region in cell.bounding_regions:
            print(
                "...content on page {} is within bounding box '{}'".format(
                    region.page_number,
                    format_bounding_box(region.bounding_box),
                )
            )

print("----------------------------------------")
```

> Analyzing general prebuilt document types with `3.2.x`:

Reviewer 2: I feel like we generally go 3.1.x and then show 3.2.x but this example doesn't have its 3.1.x counterpart. I know prebuilt-document didn't exist in 3.1.x, but I think we should include a sentence about how this should replace training without labels.

```python
with open(path_to_sample_documents, "rb") as f:
    poller = document_analysis_client.begin_analyze_document(
        "prebuilt-document", document=f
    )
result = poller.result()

> for idx, style in enumerate(result.styles):

Reviewer 2: not using idx here.

    print("...within '{}' bounding regions".format(format_bounding_region(entity.bounding_regions)))
    print("...with confidence {}\n".format(entity.confidence))

print("----Key-value pairs found in document----")
for kv_pair in result.key_value_pairs:
    if kv_pair.key:
        print(
                "Key '{}' found within '{}' bounding regions".format(
                    kv_pair.key.content,
                    format_bounding_region(kv_pair.key.bounding_regions),
                )
            )
    if kv_pair.value:
        print(
                "Value '{}' found within '{}' bounding regions\n".format(
                    kv_pair.value.content,
                    format_bounding_region(kv_pair.value.bounding_regions),
                )
            )
print("----------------------------------------")
```

> NOTE: All of these samples also work with `begin_analyze_document_from_url` when providing a valid URL to the document.

### Analyzing a custom model

Differences between the versions:
- Analyzing a custom model with `DocumentAnalysisClient` uses the general `begin_analyze_document` and `begin_analyze_document_from_url` methods.

> - In order to analyze a custom model with `FormRecognizerClient` the `begin_recognize_custom_models` and its corresponding URL methods is used.

Reviewer 2: nit

```suggestion
- In order to analyze a custom model with `FormRecognizerClient` the `begin_recognize_custom_models` and its corresponding URL methods are used.
```

for page in result.pages:
    print("\nLines found on page {}".format(page.page_number))
    for line in page.lines:
        print("...Line '{}'".format(line.content))
    for word in page.words:
        print(
            "...Word '{}' has a confidence of {}".format(
                word.content, word.confidence
            )
        )
    if page.selection_marks:
        print("\nSelection marks found on page {}".format(page.page_number))
        for selection_mark in page.selection_marks:
            print(
                "...Selection mark is '{}' and has a confidence of {}".format(
                    selection_mark.state, selection_mark.confidence
                )
            )

for i, table in enumerate(result.tables):
    print("\nTable {} can be found on page:".format(i + 1))
    for region in table.bounding_regions:
        print("...{}".format(i + 1, region.page_number))
    for cell in table.cells:
        print(
            "...Cell[{}][{}] has text '{}'".format(
                cell.row_index, cell.column_index, cell.content
            )
        )
print("-----------------------------------")
```

### Training a custom model

Differences between the versions:

> - Files for building a new model for version `3.2.x` can be created using the labeling tool found [here](fr-labeling-tool).

Reviewer: ```suggestion
- Files for building a new model for version `3.2.x` can be created using the labeling tool found [here][fr-labeling-tool].
```

> - In version `3.1.x` the `use_training_labels` kwarg was used to indicate whether to use labeled data was when creating the custom model.

Reviewer: ```suggestion
- In version `3.1.x` the `use_training_labels` kwarg was used to indicate whether to use labeled data when creating the custom model.
```

- In version `3.2.x` the `use_training_labels` kwargs is not supported since training must be carried out with labeled training documents. In order to extract key-value pairs from a document, please refer to the prebuilt model "prebuilt-document" which extracts entities, key-value pairs, and layout from a document. 

> - When using the latest service API version `2021-09-30-preview` models no longer include submodels, instead a model can analyze different document types.

Reviewer: ```suggestion
- When using the latest service API version `2021-09-30-preview` models no longer include submodels. Instead a model can analyze different document types.
```

Author: I think I'll keep this one as is since the second part of the sentence needs the first part for context.

> - In version `3.1.x`, when training without labels the service would returna model that recognized key-value pairs in the training data. When using the library with version `3.2.x`, users can instead use the prebuilt model `prebuilt-document` to get this information from documents without the need to train a new model. `prebuilt-document` returns key-value pairs, entities, text layout, among other useful information about the documents passed in for analysis.

Reviewer: this feels like a repetition of the last 2 points. Consider removing it

Reviewer 2: I don't think we should say 3.1.x returned key-value pairs, these were exposed as RecognizedForm.fields.

Author: So I actually removed this part in favor of the previous explanation like Mari suggested since I do mention using "prebuilt-document" in one of the previous points.

> - When buildong or composing new models users can now assign their own model ids and specify a description.

Reviewer: ```suggestion
- When building or composing new models users can now assign their own model ids and specify a description.
```

# Guide for migrating azure-ai-formrecognizer to version 3.2.x from versions 3.1.x and below

This guide is intended to assist in the migration to `azure-ai-formrecognizer (3.2.x)` from versions `3.1.x` and below. It will focus on side-by-side comparisons for similar operations between versions. Please note that version `3.2.0b1` will be used for comparison with `3.1.2`.

Familiarity with `azure-ai-formrecognizer (3.1.x and below)` package is assumed. For those new to the Azure Form Recognizer client library for Python please refer to the [README][readme] rather than this guide.

## Table of Contents
- [Migration benefits](#migration-benefits)
- [Important changes](#important-changes)
    - [Client usage](#client-usage)
    - [Analyzing document](#analyzing-documents)

>     - [Analyzing a custom model](#analyzing-a-custom-model)

Reviewer 3: Analyzing a custom model => Analyzing a document with a custom model

>     - [Training a custom model](#training-a-custom-model)

Reviewer 3: Building a custom model?

Author: I left it as training since the guide is catering more towards people coming from the previous version. I commented on this here: https://github.com/Azure/azure-sdk-for-python/pull/20960#discussion_r719703484

    - [Manage models](#manage-models)
- [Additional samples](#additional-samples)

## Migration benefits

> A natural question to ask when considering whether or not to adopt a new version of the library is what the benefits of doing so would be. As Azure Form Recognizer has matured and been embraced by a more diverse group of developers, we have been focused on learning the patterns and practices to best support developer productivity and add value to our customers.

Reviewer 3: nit: whether or not => whether

Please refer to the [README][readme] for more information on these new clients.

## Important changes

### Client usage

We continue to support API key and AAD authentication methods when creating the clients. Below are the differences between the two versions:

- In `3.2.x`, we have added `DocumentAnalysisClient` and `DocumentModelAdministrationClient` which support API version `2021-09-30-preview` and later.
- `FormRecognizerClient` and `FormTrainingClient` will raise an error if called with an API version of `2021-09-30-preview` and later. 
- In `DocumentAnalysisClient` all prebuilt model methods along with custom model, layout, and a prebuilt document analysis model are unified into two methods called
`begin_analyze_document` and `begin_analyze_document_from_url`.
- In `FormRecognizerClient` there are two methods (a stream and URL method) for each of the prebuilt models supported by the service. This results in two methods for business card, receipt, identity document, and invoice models, along with a pair of methods for recognizing custom documents and for recognizing content/layout. 

Creating new clients in `3.1.x`:
```python

> form_recognizer_client = FormRecognizerClient(

Reviewer 2: sorry I should have left this comment before but I forgot. I think we should include the imports in these first client samples just to show that the namespace hasn't changed and that the new DAC/DMAC are imported from azure.ai.formrecognizer and azure.ai.formrecognizer.aio as well.

- In the new version of the library, the functionality of `begin_recognize_content` has been added as a prebuilt model and can be called in library version `azure-ai-formrecognizer (3.2.x)` with `begin_analyze_document` by passing in the `prebuilt-layout` model ID. Similarly, to get general prebuilt document information, such as key-value pairs, entities, and text layout, the `prebuilt-document` model ID can be used with `begin_analyze_document`.
- When calling `begin_analyze_document` and `begin_analyze_document_from_url` the returned type is an `AnalyzeResult` object, while the various methods used with `FormRecognizerClient` return a list of `RecognizedForm`.
- The `pages` keyword argument is a string with library version `azure-ai-formrecognizer (3.2.x)`. In `azure-ai-formrecognizer (3.1.x)`, `pages` was a list of strings.
- The `include_field_elements` keyword argument is not supported with the `DocumentAnalysisClient`, text details are automatically included with API version `2021-09-30-preview` and later.

> - The `reading_order` keyword argument does not exist on `begin_analyze_document` and `begin_analyze_document_from_url`. The service uses `natural` reading order to analyze data.

Reviewer 2: I think return makes more sense here since the data is returned in the natural reading order

```suggestion
- The `reading_order` keyword argument does not exist on `begin_analyze_document` and `begin_analyze_document_from_url`. The service uses `natural` reading order to return data.
```

                word.content, word.confidence
            )
        )
    if page.selection_marks:
        print("\nSelection marks found on page {}".format(page.page_number))
        for selection_mark in page.selection_marks:
            print(
                "...Selection mark is '{}' and has a confidence of {}".format(
                    selection_mark.state, selection_mark.confidence
                )
            )

for i, table in enumerate(result.tables):
    print("\nTable {} can be found on page:".format(i + 1))
    for region in table.bounding_regions:
        print("...{}".format(i + 1, region.page_number))
    for cell in table.cells:
        print(
            "...Cell[{}][{}] has text '{}'".format(
                cell.row_index, cell.column_index, cell.content
            )
        )
print("-----------------------------------")
```

### Training a custom model

Differences between the versions:
- Files for building a new model for version `3.2.x` can be created using the labeling tool found [here][fr_labeling_tool].
- In version `3.1.x` the `use_training_labels` keyword argument was used to indicate whether to use labeled data when creating the custom model.

> - In version `3.2.x` the `use_training_labels` keyword argument is not supported since training must be carried out with labeled training documents. In order to extract key-value pairs from a document, please refer to the prebuilt model "prebuilt-document" which extracts entities, key-value pairs, and layout from a document. 

Reviewer 2: sorry to bring this up again, but I don't think we should mention key-value pairs as this was not what it was referred to in v3.1.x. maybe something like?

"Additionally train without labels is now replaced with the prebuilt model "prebuilt-document"...."

    print("Document status: {}".format(doc.status))
    print("Document page count: {}".format(doc.page_count))
    print("Document errors: {}".format(doc.errors))
```

Train a custom model with `3.2.x`:
```python
document_model_admin_client = DocumentModelAdministrationClient(endpoint, AzureKeyCredential(key))
poller = document_model_admin_client.begin_build_model(
    container_sas_url, model_id="my-model-id", description="my model description"
)
model = poller.result()

print("Model ID: {}".format(model.model_id))
print("Description: {}".format(model.description))
print("Model created on: {}\n".format(model.created_on))
print("Doc types the model can recognize:")
for name, doc_type in model.doc_types.items():
    print("\nDoc Type: '{}' which has the following fields:".format(name))
    for field_name, confidence in doc_type.field_confidence.items():
        print("Field: '{}' has confidence score {}".format(field_name, confidence))
```

### Manage models

Differences between the versions:
- When using API version `2021-09-30-preview` and later models no longer include submodels, instead a model can analyze different document types.

> - When building or composing new models users can now assign their own model IDs and specify a description.

Reviewer 2: and copy

> - In version `3.2.x` of the library, only models that succeeded can be retreived from get and list calls. In version `3.1.x` of the library, models that had not succeeded were still created and had to be deleted.

Reviewer 2: something like?

```suggestion
- In version `3.2.x` of the library, only models that build successfully can be retrieved from the get and list models calls. Unsuccessful model operations can be viewed with the get and list operations methods (note that document model operation data persists for only 24 hours). In version `3.1.x` of the library, models that had not succeeded were still created, had to be deleted by the user, and were returned in the list models response.
```

There are many benefits to using the new design of the `azure-ai-formrecognizer (3.2.x)` library. This new version of the library introduces two new clients `DocumentAnalysisClient` and the `DocumentModelAdministrationClient` with unified methods for analyzing documents and provides support for the new features added by the service in API version `2021-09-30-preview` and later.

New features provided by the `DocumentAnalysisClient` include one consolidated method for analyzing document layout, a general prebuilt document model type, along with the same prebuilt models that were included previously (receipts, invoices, business cards, identity documents), and custom models. Moreover, the models introduced in the latest version of the library, such as `AnalyzeResult`, remove hierarchical dependencies between document elements and move them to a more top level and easily accessible position. The service has further improved how to define where elements are located on documents by moving towards `BoundingRegion` definitions allowing for cross-page elements. Document element fields are returned with more information, such as content and spans. 

> When using the `DocumentModelAdministrationClient` to build, compose, or copy models, users can now assign their own model IDs and specify a description. Listing models on the administration client now includes both prebuilt and custom models. When using `get_model()`, users can get the field schema (field names and types that the model can extract) for the model they specified, including for prebuilt models. This client also provides functions for getting information from model operations.

Reviewer 3: For the last line,
>This client also provides functions for getting information from model operations.

Do we need to mention anything about past 24 hours valid operation ID ?

Author: We do on line 643, I took this part as more of an intro/summary. The details are in the bullets in each section.
